crates/extrinsics/src/contract_info.rs

```diff
@@ -32,45 +32,128 @@ use scale::Decode;
+    struct DepositAccount {
+        deposit_account: AccountId32,
+    }
+
+    let account = contract_info.as_type::<DepositAccount>()?;
+    Ok(account.deposit_account)
+}
+
 /// Fetch the contract info from the storage using the provided client.
 pub async fn fetch_contract_info(
     contract: &AccountId32,
     rpc: &LegacyRpcMethods<DefaultConfig>,
     client: &Client,
-) -> Result<Option<ContractInfo>> {
-    let info_contract_call = api::storage().contracts().contract_info_of(contract);
-
+) -> Result<ContractInfo> {
     let best_block = get_best_block(rpc).await?;
 
-    let contract_info_of = client
+    let contract_info_address = dynamic(
+        "Contracts",
+        "ContractInfoOf",
+        vec![Value::from_bytes(contract)],
+    );
+    let contract_info = client
         .storage()
         .at(best_block)
-        .fetch(&info_contract_call)
-        .await?;
+        .fetch(&contract_info_address)
+        .await?
+        .ok_or_else(|| {
+            anyhow!(
+                "No contract information was found for account id {}",
+                contract
+            )
+        })?;
+    #[derive(DecodeAsType, Debug)]
+    #[decode_as_type(crate_path = "subxt::ext::scale_decode")]
+    pub struct BoundedVec<T>(pub ::std::vec::Vec<T>);
+    #[derive(DecodeAsType, Debug)]
+    #[decode_as_type(crate_path = "subxt::ext::scale_decode")]
+    struct ContractInfoOf {
+        trie_id: BoundedVec<u8>,
+        code_hash: CodeHash,
+        storage_items: u32,
+        storage_item_deposit: Balance,
+    }
 
-    match contract_info_of {
-        Some(info_result) => {
-            let convert_trie_id = hex::encode(info_result.trie_id.0);
-            Ok(Some(ContractInfo {
-                trie_id: convert_trie_id,
-                code_hash: info_result.code_hash,
-                storage_items: info_result.storage_items,
-                storage_item_deposit: info_result.storage_item_deposit,
-            }))
+    // Pallet-contracts [>=10, <15] store the contract's deposit as a free balance
+    // in a secondary account (deposit account). Other versions store it as
+    // reserved balance on the main contract's account. If the
+    // `deposit_account` field is present in a contract info structure,
+    // the contract's deposit is in this account.
+    let total_balance: Balance = match get_deposit_account_id(&contract_info) {
+        Ok(deposit_account) => {
+            get_account_balance(&deposit_account, rpc, client)
+                .await?
+                .free
         }
-        None => Ok(None),
-    }
+        Err(_) => get_account_balance(contract, rpc, client).await?.reserved,
+    };
+
+    let info = contract_info.as_type::<ContractInfoOf>()?;
+    Ok(ContractInfo {
+        trie_id: hex::encode(info.trie_id.0),
+        code_hash: info.code_hash,
+        storage_items: info.storage_items,
+        storage_items_deposit: info.storage_item_deposit,
+        storage_total_deposit: total_balance,
+    })
```

Comment:
I actually meant to extract and test this whole block, to see that it works for both versions of the metadata.

Reply:
Suggestion for this here: https://github.com/paritytech/cargo-contract/pull/1400/files

Reply:
Done